Autodesk/synthesis
# WPILibBrain Mapping

## Changes:
- Added mapping of PWM and CAN groups
  - Modals to select initialized PWM ports and drivers
- Added mirabuf joint instance info to drivers so the config UI can display joint names along with driver types to better differentiate them when selecting

## Testing:
1. [Run simulator](#running_simulator)
1. Create PWM group (CAN not within the scope of this PR)
    1. Ensure WS data is there and only the initialized PWM port (0 in Java sample code) shows up on the left
    1. Ensure driver type and joint names show up on the right
1. When in disconnected or disabled mode, console should print `0 0`, meaning port `0` and speed `0`
1. Enable autonomous mode in the simulator and ensure it now prints `0 0.5`
1. Enable teleop mode and ensure it prints `0 0.25`

> [!NOTE]
> There is commented out PWM driving code in `WPILibBrain.ts` in the `PWMGroup` class `Update` method. You can uncomment it to make the robot actually move instead of just printing the speed to the console.

### <a name="running_simulator"></a>Running the simulator
1. Navigate to the `simulation` directory in the root of the synthesis project
2. Navigate into `SyntheSimJava
3. Run `./gradlew build && ./gradlew publishToMavenLocal`
4. Navigate to `simulation/samples/JavaSample`
5. Install [WPILib](https://docs.wpilib.org/en/stable/docs/zero-to-robot/step-2/wpilib-setup.html) utilities if you don't have them
6. Open the `JavaSample` project in WPILib VSCode
7. Install Phoenix v6 and REVLib vendor libraries
    1. `https://maven.ctr-electronics.com/release/com/ctre/phoenix6/latest/Phoenix6-frc2024-latest.json`
    2. `https://software-metadata.revrobotics.com/REVLib-2024.json`
8. Run `./gradlew build && ./gradlew simulateJava`

> [!NOTE]
> See the [SyntheSimJava README](https://github.com/Autodesk/synthesis/tree/jwrigh/1746/wpilibbrain-mapping/simulation/SyntheSimJava/README.md) for documentation on the library.

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1746)
